runtime/compiler/control/CompilationThread.cpp

```diff
@@ -3470,71 +3456,61 @@ void TR::CompilationInfo::stopCompilationThreads()
-            curCompThreadInfoPT->getCompThreadMonitor()->enter();
-            curCompThreadInfoPT->getCompThreadMonitor()->notifyAll();
-            curCompThreadInfoPT->getCompThreadMonitor()->exit();
-            break;
+   TR_ASSERT_FATAL(getNumCompThreadsActive() == 0, "All threads must be inactive at this point\n");
 
-         case COMPTHREAD_ACTIVE:
-         case COMPTHREAD_SIGNAL_WAIT:
-         case COMPTHREAD_WAITING:
-            curCompThreadInfoPT->setCompilationThreadState(COMPTHREAD_SIGNAL_TERMINATE);
-            // Only decrement the number of active threads if we are not a diagnostic
-            // thread
-            if (!(curCompThreadInfoPT->isDiagnosticThread()))
-               decNumCompThreadsActive();
-            break;
+   purgeMethodQueue(compilationSuspended);
 
-         case COMPTHREAD_SIGNAL_SUSPEND:
-            curCompThreadInfoPT->setCompilationThreadState(COMPTHREAD_SIGNAL_TERMINATE);
-            break;
+   // Cycle though all non-diagnostic threads and wait for them to terminate. At this point it is possible that a
+   // compilation thread has crashed and is going through the JitDump process. The reason we skipped terminating the
+   // diagnostic threads in the above loop is because the JitDump logic will activate the diagnostic thread to generate
+   // the JitDump, so the diagnostic thread must not be in a terminated state at that point.
+   for (uint8_t i = 0; i < getNumTotalCompilationThreads(); i++)
+      {
+      TR::CompilationInfoPerThread *curCompThreadInfoPT = _arrayOfCompilationInfoPerThread[i];
 
-         case COMPTHREAD_SIGNAL_TERMINATE:
-         case COMPTHREAD_STOPPING:
-         case COMPTHREAD_STOPPED:
-            // weird case; we should not do anything
-            break;
+      if (!curCompThreadInfoPT->isDiagnosticThread())
+         {
+         while (curCompThreadInfoPT->getCompilationThreadState() != COMPTHREAD_STOPPED)
+            {
+            getCompilationMonitor()->notifyAll();
+            waitOnCompMonitor(vmThread);
+            }
+         }
+      }
 
-         case COMPTHREAD_UNINITIALIZED:
-            // Compilation thread did not have time to become fully initialized
-            curCompThreadInfoPT->setCompilationThreadState(COMPTHREAD_SIGNAL_TERMINATE);
-            break;
+   // Now that all compilation threads have terminated, we are sure none of them have crashed, or they have finished
+   // generating a JitDump. Only at this point can we terminate the diagnostic threads.
+   for (uint8_t i = 0; i < getNumTotalCompilationThreads(); i++)
+      {
+      TR::CompilationInfoPerThread *curCompThreadInfoPT = _arrayOfCompilationInfoPerThread[i];
 
-         default:
-            TR_ASSERT(false, "No other comp thread state possible here\n");
-         } // end switch
+      if (curCompThreadInfoPT->isDiagnosticThread())
+         stopCompilationThread(curCompThreadInfoPT);
```

Comment:
We are not sure. The `stopCompilationThread` API will signal the thread to terminate at the next yield point, however (work yet to come in #11860) diagnostic thread compilations will be uninterruptible, meaning we will block waiting on the monitor a few lines below on line 3509 in the next loop.

This ensures that the diagnostic thread will finish it's compilations before the code being reviewed here has stopped the thread.

Reply:
The above is for crashes in application threads. For crashes in compilation threads, the loop preceding this will block because all non-diagnostic compilation threads need to terminate before we terminate the diagnostic thread. But a non-diagnostic thread has crashed, and is currently active, so it will _not_ terminate, because it will not reach the state processing loop in `TR::CompilationInfoPerThread::run()`. Thus we will never execute this code for crashes in a compilation thread.

This is because we block in the loop preceding this code being reviewed, and once the JitDump has been generated we return to finish running the dump generation process on the crashed thread. Once the dumps have been generated the JVM will `exit(-1)`.

Reply:
I agree with the explanation and yes, it would be helpful to add that as a comment in the codebase.

Reply:
Added more comments about this in f9ff0c5. Hopefully future readers will understand.